src/run_program.rs

```diff
@@ -326,9 +349,11 @@ where
                 None => break,
             };
             cost += match op {
-                Operation::Apply => self.apply_op()?,
+                Operation::Apply => {
+                    augment_cost_errors(self.apply_op(max_cost - cost), &max_cost_ptr)?
```

Comment:
the `augment_cost_errors()` is the moral equivalence of catching `CostError` exceptions, setting the `max_cost` field and rethrowing.